### PR TITLE
Scheduler, make stale DAG deactivation threshold configurable instead of using dag processing timeout

### PR DESCRIPTION
**Pull request by michaelmicheal**
When we deactivate stale DAGs, we add a buffer to the last parsed time of a DAG when comparing it to the last parsed time of its DAG file to account for DAG files that take a long time to process. 
```python

        for dag in dags_parsed:
            # The largest valid difference between a DagFileStat's last_finished_time and a DAG's
            # last_parsed_time is _processor_timeout. Longer than that indicates that the DAG is
            # no longer present in the file.
            if (
                dag.fileloc in last_parsed
                and (dag.last_parsed_time + processor_timeout) < last_parsed[dag.fileloc]
            ):
                cls.logger().info("DAG %s is missing and will be deactivated.", dag.dag_id)
                to_deactivate.add(dag.dag_id)

```
The upper bound on this delay is indeed the processor timeout, but practically speaking, this is unnecessary when we configure a long timeout. This can result in a large delay in stale DAG deactivation. I think this buffer should be configurable.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -1045,8 +1045,16 @@ min_file_process_interval = 30
...
+# DAGs (DAGs which are no longer present in the expected files). The reason why we need
+# this threshold is to account for the time between when the file is parsed and when the
+# DAG is loaded. The absolute maximum that this could take is `dag_file_processor_timeout`,
+# but when you have a long timeout configured, it results in a significant delay in the
+# deactivation of stale dags.
+stale_dag_threshold = 30
+
 # How often (in seconds) to scan the DAGs directory for new files. Default to 5 minutes.
-dag_dir_list_interval = 300
+dag_dir_list_interval = 10
```
Comment (ephraimbuddy) [inline airflow/config_templates/default_airflow.cfg]: Why this change?

File: airflow/dag_processing/manager.py [hunk 2]
```diff
@@ -433,6 +433,8 @@ def __init__(
         self.last_deactivate_stale_dags_time = timezone.make_aware(datetime.fromtimestamp(0))
         # How often to check for DAGs which are no longer in files
         self.parsing_cleanup_interval = conf.getint("scheduler", "parsing_cleanup_interval")
+        # How long to wait for a DAG to be reparsed after it's file has been parsed before disabling
```
Comment (ephraimbuddy) [inline airflow/dag_processing/manager.py]:
```suggestion
        # How long to wait for a DAG to be reparsed after its file has been parsed before disabling
```

File: airflow/dag_processing/manager.py [hunk 3]
```diff
@@ -523,13 +525,16 @@ def deactivate_stale_dags(
             query = query.filter(DagModel.processor_subdir == dag_directory)
         dags_parsed = query.all()
 
+
+
         for dag in dags_parsed:
             # The largest valid difference between a DagFileStat's last_finished_time and a DAG's
-            # last_parsed_time is _processor_timeout. Longer than that indicates that the DAG is
-            # no longer present in the file.
+            # last_parsed_time is thg processor_timeout. Longer than that indicates that the DAG is
```
Comment (ephraimbuddy) [inline airflow/dag_processing/manager.py]:
```suggestion
            # last_parsed_time is the processor_timeout. Longer than that indicates that the DAG is
```

File: airflow/config_templates/config.yml [hunk 4]
```diff
@@ -2049,6 +2049,18 @@ scheduler:
...
+        How long (in seconds) to wait after we've reparsed a DAG file before deactivating stale
+        DAGs (DAGs which are no longer present in the expected files). The reason why we need
+        this threshold is to account for the time between when the file is parsed and when the
+        DAG is loaded. The absolute maximum that this could take is `dag_file_processor_timeout`,
+        but when you have a long timeout configured, it results in a significant delay in the
+        deactivation of stale dags.
+      version_added: 2.6.0
+      type: integer
+      example: ~
+      default: "30"
```
Comment (ephraimbuddy) [inline airflow/config_templates/config.yml]: Should we keep this the same default as `dag_file_processor_timeout`? That way, there will be no change in behaviour

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (michaelmicheal) [reply airflow/config_templates/default_airflow.cfg]: Sorry, accident

**Review by ephraimbuddy — COMMENTED** [review]
Can we have a test with different values for the threshold? This would gaurd against removing  the config without a _fight_

**Comment by michaelmicheal** [comment]
> Can we have a test with different values for the threshold? This would gaurd against removing the config without a _fight_
Is there an example of something like this for other configurations?

**Comment by ephraimbuddy** [comment]
> > Can we have a test with different values for the threshold? This would gaurd against removing the config without a _fight_
> 
> Is there an example of something like this for other configurations?

Yep. https://github.com/apache/airflow/blob/02a2efeae409bddcfedafe273fffc353595815cc/tests/jobs/test_scheduler_job.py#L527

**Review by ephraimbuddy — APPROVED** [review]

File: airflow/config_templates/config.yml [hunk 5]
```diff
@@ -2049,6 +2049,18 @@ scheduler:
       type: integer
       example: ~
       default: "60"
+    stale_dag_threshold:
+      description: |
+        How long (in seconds) to wait after we've reparsed a DAG file before deactivating stale
```
Comment (Taragolis) [inline airflow/config_templates/config.yml]:
```suggestion
        How long (in seconds) to wait after we have reparsed a DAG file before deactivating stale
```

This should fix, spellchecker doesn't like some words with `'`

```console
Incorrect Spelling: 'reparsed'
Line with Error: 'How long (in seconds) to wait after we've reparsed a DAG file before deactivating stale'
Line Number: 6812
```

**Comment by Taragolis** [comment]
I would recommend to use `pre-commit` to fix most of the statics checks, see: [Static code checks](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#id4)

File: airflow/config_templates/config.yml (hunk 5 above)
Comment (Taragolis) [reply airflow/config_templates/config.yml]: You could also check it in [Breeze](https://github.com/apache/airflow/blob/main/BREEZE.rst#building-the-documentation) locally

File: airflow/dag_processing/manager.py [hunk 6]
```diff
@@ -523,13 +525,17 @@ def deactivate_stale_dags(
             query = query.filter(DagModel.processor_subdir == dag_directory)
         dags_parsed = query.all()
 
+        processor_timeout_seconds: int = conf.getint("core", "dag_file_processor_timeout")
+
+
```
Comment (ephraimbuddy) [inline airflow/dag_processing/manager.py]:
```suggestion
```

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
Doesn't seem that the `processor_timeout_seconds` was used anywhere

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Needs fixing tests :(

**Comment by potiuk** [comment]
Rebased to account for workarounded issue with boto3. But seems the issue has been fixed @ephraimbuddy

**Comment by michaelmicheal** [comment]
> Rebased to account for workarounded issue with boto3. But seems the issue has been fixed @ephraimbuddy

Thanks is this good to merge?

**Comment by potiuk** [comment]
> > Rebased to account for workarounded issue with boto3. But seems the issue has been fixed @ephraimbuddy
> 
> Thanks is this good to merge?

Waiitng for @ephraimbuddy to remove his "change request" . I cannot merge it without it.

**Review by ephraimbuddy — APPROVED** [review]
